Pullcoronalplane

AnnaEKnight opened the pull request:
This should allow for more functionality and let a user specify what plane they would like to pull in terms of both what plane it is constant in (for example an elevational plane would be constant in elevation but vary in axial and lateral), and also allow for specifying exactly which plane is pulled (ex. plane_pos = 2 with plane_orientation = 2 means take the plane that is constant at axial position of 2, that varies in lateral and elevational)

This pull request works, and does not break the existing tests-- ideally i would like to write more tests for this in the future.

AnnaEKnight committed:
trying to update so that create res sim can now take in a plane direction instead of just elevational plane.

AnnaEKnight committed:
syntax clean up

AnnaEKnight committed:
updating create_res_sim to actually pull the correct plane hopefully as previously did not handle indexing properly, also testing if i can commit from inside a venv on jedi with spyder through fastX

AnnaEKnight committed:
making progress towards fixing the save format to save the correct axes depending on what plane is extracted, but something is broken

AnnaEKnight committed:
okay i think now both pulling a coronal plane and pulling a standard elevational plane with the old notation is working

AnnaEKnight committed:
getting rid of local temp files

AnnaEKnight committed:
adding a bunch of print statements to understand what the values in the planes are for writing more robust tests  now that test_extract_image can take multiple directions. annoying but the only way i could think to do this

AnnaEKnight committed:
Merge branch 'pullcoronalplane' of https://github.com/mlp6/fem into pullcoronalplane

AnnaEKnight committed:
this one might actually run

mlp6 commented on examples/gauss_qsym_pml/runSaveCoronalExamples.py:
```diff
@@ -0,0 +1,64 @@
+create_res_sim.extract3Darfidata(dynadeck=DYNADECK,
+                                 dispout="disp.dat",
+                                 ressim="res_sim.pvd")
+##create_res_sim.run(dynadeck=DYNADECK,
```
You can blast away these lines that are commented out.

AnnaEKnight replied:
I think I can just blast this whole file actually? it was more for my own test casing than anything that needs to be preserved

mlp6 commented on fem/post/create_res_sim.py:
```diff
@@ -182,12 +187,18 @@ def __read_cli():
 
 
-def extract_image_plane(snic, axes, ele_pos):
+def extract_image_plane(snic, axes, plane_pos, direction = 0):
```
We should check (with a test) if the old syntax using the `ele_pos` keyword argument will still work if someone has old code that doesn't utilize these new keyword arguments.  If not, we'll increment up the major version number of the repo to reflect the API change.

mlp6 commented on fem/post/create_res_sim.py:
```diff
@@ -199,18 +210,62 @@ def extract_image_plane(snic, axes, ele_pos):
+    plane0 = np.min(np.where(axes[direction]>=plane_pos))
+
+    if direction == 0:
+        image_plane = np.squeeze(snic['id'][plane0,:,:]).astype(int)
```
Is the only difference in each `elif` block the definition of `image_plane`?  If so, then all of those `print()` statements can be reduced to something outside that conditional logic (i.e., not repeated).  Also, instead of `print()` statements, the package uses the `logging` package; this will both display output to the terminal *and* capture it in a log file so that you can go back and audit what was done.  Check out replacing these prints with `logging.info()` statements.

AnnaEKnight replied:
sorry these print statements were when i thought i was writing test cases for the other plane cases and wanted to know for example where image_plane[1][1] was for the coronal plane-- i can just blast them away

AnnaEKnight committed:
getting rid of print statements that were unnecessary

AnnaEKnight committed:
this file was not tracked for                some reason

AnnaEKnight committed:
adding tests for orthogonal cases for test_create_res_sim.py

AnnaEKnight commented:
@mlp6 i think this is good to go now

mlp6 commented:
I am getting a failed unit test:
```
==================================================================== FAILURES ====================================================================
____________________________________________________________ test_extract_image_plane ____________________________________________________________
[gw2] linux -- Python 3.9.4 /home/mlp6/Projects/fem/.venv/bin/python

sorted_nodes = array([[[(   1, -1. , 0. , -1. ), ( 122, -1. , 0. , -0.9),
         ( 243, -1. , 0. , -0.8), ..., ( 969, -1. , 0. , -0...0,  0. , 1. , -0.1), (1331,  0. , 1. ,  0. )]]],
      dtype=[('id', '<i4'), ('x', '<f4'), ('y', '<f4'), ('z', '<f4')])
axes = [array([-1. , -0.9, -0.8, -0.7, -0.6, -0.5, -0.4, -0.3, -0.2, -0.1,  0. ],
      dtype=float32), array([0. , 0.1, 0.2,...,
      dtype=float32), array([-1. , -0.9, -0.8, -0.7, -0.6, -0.5, -0.4, -0.3, -0.2, -0.1,  0. ],
      dtype=float32)]

    def test_extract_image_plane(sorted_nodes, axes):
        from fem.post.create_res_sim import extract_image_plane
    
        image_plane = extract_image_plane(sorted_nodes, axes, 0.0)
    
        assert image_plane.shape == (11, 11)
        assert image_plane[0][0] == 11
        assert image_plane[0][-1] == 1221
        assert image_plane[-1][0] == 121
        assert image_plane[-1][-1] == 1331
    
        image_plane = extract_image_plane(sorted_nodes, axes, plane_pos = 0, direction =1)
    
        assert image_plane.shape == (11,11)
>       assert image_plane[0][0] == 11
E       assert 1 == 11
E         +1
E         -11

tests/test_create_res_sim.py:30: AssertionError
============================================================ short test summary info =============================================================
FAILED tests/test_create_res_sim.py::test_extract_image_plane - assert 1 == 11
========================================================= 1 failed, 49 passed in 34.87s ==========================================================
```
I'm not sure why this test didn't run correctly during the push / pull request trigger.  I'll look into that while you look into the issue above.

mlp6 committed:
ADD: flake8 package for linting

mlp6 committed:
EDIT: GitHub Python workflow for pytest auto-discovery

The actual unit tests weren't running with GitHub actions... so
"passing" was just things running, but not actually testing.  Trying to
explicitly specify the tests/ directory.

mlp6 commented:
@AnnaEKnight I fixed the Python-specific testing to make sure the tests are actually run.

mlp6 commented on fem/post/create_res_sim.py:
```diff
@@ -199,18 +220,26 @@ def extract_image_plane(snic, axes, ele_pos):
       ele_pos: elevation position for plane of interest
```
@AnnaEKnight docstring inputs need to be updated for new keyword arguments and details about what the valid input values mean

AnnaEKnight committed:
fixing the doc strings and trying to fix the failing tests- hopefully they actually fail if they are broken this tie

AnnaEKnight committed:
orientation 1 passed on the last commit now need to get orientation 2 to pass

AnnaEKnight commented:
okay i think it ACTUALLY ran this time

mlp6 commented on fem/post/create_res_sim.py:
```diff
@@ -196,21 +217,32 @@ def extract_image_plane(snic, axes, ele_pos):
 
-def save_res_sim(resfile, arfidata, axes, t, axis_scale=(-10, 10, -10)):
+
+def save_res_sim(resfile, arfidata, axes, t, axis_scale=(-10, 10, -10), plane_pos = 0, plane_orientation = 0):
```
@AnnaEKnight need to also update this docstring

AnnaEKnight committed:
fixing doc strings on save_res_sim

mlp6 committed:
ADD: legacy ele_pos kwarg input test

mlp6 committed:
RM: ele_pos kwargs; ADD: missing raised exception

* ele_pos kwarg has been deprecated and replaced with Anna's more robust
  arbitrary plane extraction.  It is still accepted, but shouldn't be
  represented in the promoted workflows / be redundant with the new
  kwargs.

* Added a missing raised exception to the docstring for savepvd().

mlp6 committed:
VERSION UPDATE: v8.1.4 -> v8.2.0

mlp6 committed:
REFACTOR: more robust syntax checking and exception raising

* ele_pos as a kwarg was being touted as deprecated but still valid, but
  it actually would raise a TypeError exception.  This raised exception
  is now tested.

* plane0 internal variable refactored to plane since the "0" did not
  convey any meaning and was a legacy variable name from when the
  elevation-only plane was also limited to the elevation = 0 plane.

* Add explicity type hinting that plane_pos should be a float, and
  direction an int.  Additionally, plane_pos now passed as 0.0 instead
  of 0 to make this clear and consistent with the type hint.

* Check that direction is in bounds of valid values are raise a
  ValueError exception if it is not in bounds.  This check is now done
  before the conditional tree, since direction is called as an index to
  the array before than conditional tree is reached, and an
  out-of-bounds values will raise an exception and not gracefully exit.

mlp6 commented:
@AnnaEKnight take a look at the commits that I made today (I think well summarized in the commit messages) to make sure you agree w/ my changes.
